pallets/phala/src/compute/vault.rs

```diff
@@ -0,0 +1,511 @@
+			target: T::AccountId,
+			shares: BalanceOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			let mut pool_info = ensure_vault::<T>(vault_pid)?;
+			ensure!(
+				who == pool_info.basepool.owner,
+				Error::<T>::UnauthorizedPoolOwner
+			);
+			ensure!(
+				pool_info.owner_shares >= shares,
+				Error::<T>::InvaildWithdrawSharesAmount
+			);
+			ensure!(shares > Zero::zero(), Error::<T>::NoRewardToClaim);
+			let _nft_id = base_pool::Pallet::<T>::mint_nft(pool_info.basepool.cid, target, shares, vault_pid, PoolType::Vault)?;
+			pool_info.owner_shares -= shares;
+			base_pool::pallet::Pools::<T>::insert(vault_pid, PoolProxy::Vault(pool_info));
+			Self::deposit_event(Event::<T>::OwnerSharesClaimed {
+				pid: vault_pid,
+				user: who,
+				shares,
+			});
+
+			Ok(())
+		}
+
+		/// Tries to settle owner shares if the vault profits
+		///
+		/// The mechanism of issuing shares to distribute owner reward is metioned in comments of struct `Vault` in poolproxy.rs
+		///
+		/// Requires:
+		/// 1. The sender is the owner
+		#[pallet::weight(0)]
+		pub fn maybe_gain_owner_shares(origin: OriginFor<T>, vault_pid: u64) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let mut pool_info = ensure_vault::<T>(vault_pid)?;
+			// Add pool owner's reward if applicable
+			ensure!(
+				who == pool_info.basepool.owner,
+				Error::<T>::UnauthorizedPoolOwner
+			);
+			let current_price = match pool_info.basepool.share_price() {
+				Some(price) => BalanceOf::<T>::from_fixed(&price),
+				None => return Ok(()),
+			};
+			if pool_info.last_share_price_checkpoint == Zero::zero() {
+				pool_info.last_share_price_checkpoint = current_price;
+				base_pool::pallet::Pools::<T>::insert(vault_pid, PoolProxy::Vault(pool_info));
+				return Ok(());
+			}
+			if current_price <= pool_info.last_share_price_checkpoint {
+				return Ok(());
+			}
+			let delta_price = pool_info.commission.unwrap_or_default()
+				* (current_price - pool_info.last_share_price_checkpoint);
+			let new_price = current_price - delta_price;
+			let adjust_shares = bdiv(pool_info.basepool.total_value, &new_price.to_fixed())
+				- pool_info.basepool.total_shares;
+			pool_info.basepool.total_shares += adjust_shares;
+			pool_info.owner_shares += adjust_shares;
+			pool_info.last_share_price_checkpoint = new_price;
+
+			base_pool::pallet::Pools::<T>::insert(vault_pid, PoolProxy::Vault(pool_info));
+			Self::deposit_event(Event::<T>::OwnerSharesGained {
+				pid: vault_pid,
+				shares: adjust_shares,
+				checkout_price: new_price,
+			});
+
+			Ok(())
+		}
+
+		/// Let any user to launch a vault withdraw. Then check if the vault need to be forced withdraw all its contributions.
+		///
+		/// If the shutdown condition is met, all shares owned by the vault will be forced withdraw.
+		/// Note: This function doesn't guarantee no-op when there's error.
+		#[pallet::weight(0)]
+		#[frame_support::transactional]
+		pub fn check_and_maybe_force_withdraw(
```

Comment:
No need to change now, but it lack a lot of comments to make it readable. Let's leave the message here and add it later on.

Reply:
done

Reply:
Please leave it open for future TODO. It's not changed in this PR

Reply:
done